### ISSUE TITLE
Use Dependency-specific Basename as Prefix for Exported Dependencies in Frameworks

### ISSUE THREAD

**Issue by ankushg** [JetBrains/kotlin-native#3461]
I'm requesting a feature to **honor dependency-specific basenames as prefixes, even when the dependency is exported in a different module's framework**.

---

We are working on a multiplatform project with several modules (e.g., `moduleA`, and `moduleB`). 

In `moduleA` and `moduleB`, we declare a `basename` for the individual framework artifacts. This works fine in a vacuum.

In order to consume the modules from iOS however, we have a third module (called `ios-framework`). `ios-framework` dependends on `moduleA` and `moduleB`. 

[Following the docs for exporting dependencies in frameworks](https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html#building-final-native-binaries), we declare `export` declarations for the two modules, and also have `transitiveExport = true`.

In the final artifact, the exported classes from `moduleA` and `moduleB` are prefixed with the `basename` from `ios-framework`, instead of the `basename`s from their own modules.

---

This is _particularly_ problematic when we have **classes with the same name** in `moduleA` and `moduleB` that both need to be used in Objective-C or Swift.

If I declare 
```kotlin
data class User(val userId: Long, val username: String)
``` 
in `moduleA` and a different 
```kotlin
data class User(val userId: Long, val birthDate: Long)
``` 
in `moduleB`, one of them gets exposed in the final framework as `User` and the other gets exposed as `User_`.

Ideally, they would instead be exposed as `ModuleAUser` and `ModuleBUser`.

---

This makes things especially confusing when trying to use the resulting framework in iOS. It's unclear which `User` corresponds to functions from which module. 

Keeping class names globally unique across _multiple_ modules is  an unnecessary source of friction when you control the source for all of the modules, and downright impossible if you are consuming code that other people have written.

Ideally, the process of exporting dependencies would solve this for us.

**Comment by michalharakal** [comment]
@ankushg I'm not sure if it helps, but in similar case we are using "stitching" approach, where we are combine multiple module into 3rd one and use this as one xcode framework. This combined module just set source sets pointing to both `module A` and `module B`.
The buildscript looks like this:
```
apply plugin: 'org.jetbrains.kotlin.multiplatform'
apply plugin: 'kotlinx-serialization'
apply plugin: 'org.jetbrains.kotlin.native.cocoapods'
apply plugin: 'co.touchlab.kotlinxcodesync'
apply plugin: 'io.spring.dependency-management'

apply from: "$rootDir/versions.gradle"

archivesBaseName = 'combined'

version = "1.0"

kotlin {
    targets {
        // presets.iosArm64
        iosX64('iOS'){
            binaries {
                framework(archivesBaseName)
            }
        }
    }
    sourceSets {
        final List<String> projectsList = [
                ":common:moduleA",
                ":common:moduleB"
        ]
        commonMain {
            projectsList.forEach {
                kotlin.srcDirs += "${project(it).projectDir}/src/commonMain/kotlin"
            }
            dependencies {
                projectsList.forEach {
                    project(it).configurations.all { c ->
                        if (c.name == "commonMainApi") {
                            c.dependencies.all { dep ->
                                // note: Avoid dependencies for this project's modules like a `api project(":model")`.
                                //       Already these classes are included via `srcDir`,
                                //       so adding dependencies for projects causes compile error by double declaration.
                                if (!(dep instanceof ProjectDependency)) {
                                    api dep
                                }
                            }
                        } else if (c.name == "commonMainImplementation") {
                            c.dependencies.all { dep ->
                                if (!(dep instanceof ProjectDependency)) {
                                    implementation dep
                                }
                            }
                        }
                    }
                }

            }
        }
        iOSMain {
            final String mode = project.findProperty("XCODE_CONFIGURATION")?.toUpperCase() ?: 'DEBUG'
            projectsList.forEach {
                kotlin.srcDirs += "${project(it).projectDir}/src/iosMain/kotlin"
                if (mode == "RELEASE") {
                    kotlin.srcDirs += "${project(it).projectDir}/src/iosMain/kotlinRelease"
                } else {
                    kotlin.srcDirs += "${project(it).projectDir}/src/iosMain/kotlinDebug"
                }
            }
            dependencies {
                projectsList.forEach {
                    project(it).configurations.all { c ->
                        println("YYY configurations ${c.name}")
                        if (c.name == "iosMainApi") {
                            c.dependencies.all { dep ->
                                if (!(dep instanceof ProjectDependency)) {
                                    api dep
                                }
                            }
                        } else if (c.name == "iosMainImplementation") {
                            c.dependencies.all { dep ->
                                if (!(dep instanceof ProjectDependency)) {
                                    implementation dep
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    cocoapods {
        // Configure fields required by CocoaPods.
        summary = "Lots of Droidcon Stuff"
        homepage = "https://github.com/touchlab/DroidconKotlin"
    }
}


xcode {
    projectPath = "../../iosApp/iosApp.xcodeproj"
    target = "iosApp"
}

configurations {
    compileClasspath
}
```

This is only workaround, which confuses IDE a little bit because of multiple source sets. 

Proposed solution with a basename approach sounds good to me, but any other solution for having proper support for multiple common modules is welcommed. 

BTW, is this issue somehow related to https://github.com/JetBrains/kotlin-native/pull/3457 ?

**Comment by ankushg** [comment]
@artdfel They are unfortunately not independent. 

We've got a pretty modular codebase where smaller modules are used as building blocks for larger ones. 

In fact, we actually have a third `moduleC` that depends on `moduleA`, and we want both `moduleC` and `moduleA` to be able to be called from iOS!

I saw that this had been merged: https://github.com/JetBrains/kotlin-native/pull/3457

Would this help in the use case where `moduleA` and `moduleB` were entirely independent from each other?
